JetBrains/compose-multiplatform-core
# Use new skiko API to provide font fallbacks

TODO:
* -[x] https://github.com/JetBrains/skiko/pull/935 - should me merged first
* -[x]  Use published skiko version instead of the locally built SNAPSHOT skiko

**Description:**

This PR updates the implementation details of `FontCache` in skikoMain:
- It relies on new skiko API to ensure that manually preloaded and registered fonts can be used as a last fallback, when other fallbacks didn't work out
- Update the k/wasm demo, so it preloads the emojis font

Fixes:
- https://github.com/JetBrains/compose-multiplatform/issues/3051
- https://github.com/JetBrains/compose-multiplatform/issues/3967

With this change, it's possible to mix different characters in one Text without composing the text using AnnotatedString.
Although, when the most of glyphs are known to be from a particular font, it's more performant to specify the font explicitly when using Text(...) component or via Theme.

**Usage example:**

The fallbacks won't work automatically, the fallback fonts should be preloaded like this:

```kotlin
val fontFamilyResolver = LocalFontFamilyResolver.current

...

LaunchedEffect(Unit) {
      val notoEmojisBytes = loadEmojisFontAsBytes() // loadRes(notoColorEmoji).toByteArray()
      val fontFamily = FontFamily(listOf(Font("NotoColorEmoji", notoEmojisBytes)))
      fontFamilyResolver.preload(fontFamily)
      fontsLoaded.value = true
}
```

## Testing

- Open compose multiplatform core demo: (`./gradlew :compose:mpp:demo:wasmJsBrowserRun`)
- Go to Components -> Text Field -> Emojis
- Both text fields should show the emojis

<img width="250" alt="Screenshot 2024-06-11 at 11 22 56" src="https://github.com/JetBrains/compose-multiplatform-core/assets/7372778/2ebd1187-18bf-412c-98f3-9fb2d7c5279f">



This should be tested by QA



## Release Notes

### Fixes - Web
- Allow preloading the fallback fonts. This enables the usage of emojis and other unicode characters without manually composing the Text with AnnotatedString.
